File: visualizations/moving-charge-radiation/physics.py
r"""Liénard–Wiechert fields used by the moving-charge visualization.

The module uses dimensionless units.  ``propagation_speed`` represents
:math:`c`, and ``coulomb_constant`` represents :math:`1/(4\pi\epsilon_0)`.
"""

from __future__ import annotations

import math
from collections.abc import Sequence
from dataclasses import dataclass, replace

Vector3 = tuple[float, float, float]
EPSILON = 1e-12


@dataclass(frozen=True, slots=True)
class SourceState:
    """Position, velocity, and acceleration of the source charge."""

    t: float
    x: float
    y: float
    vx: float = 0.0
    vy: float = 0.0
    ax: float = 0.0
    ay: float = 0.0
    z: float = 0.0
    vz: float = 0.0
    az: float = 0.0


@dataclass(frozen=True, slots=True)
class FieldSample:
    """Decomposed electric and magnetic fields at an observation event."""

    electric: Vector3
    magnetic: Vector3
    velocity_electric: Vector3
    radiation_electric: Vector3
    velocity_magnetic: Vector3
    radiation_magnetic: Vector3
    source: SourceState
    retarded_time: float
    distance: float


def _lerp(left: float, right: float, amount: float) -> float:
    return left + (right - left) * amount


def _cross(left: Vector3, right: Vector3) -> Vector3:
    return (
        left[1] * right[2] - left[2] * right[1],
        left[2] * right[0] - left[0] * right[2],
        left[0] * right[1] - left[1] * right[0],
    )


def _scale(vector: Vector3, factor: float) -> Vector3:
    return vector[0] * factor, vector[1] * factor, vector[2] * factor


def _add(left: Vector3, right: Vector3) -> Vector3:
    return left[0] + right[0], left[1] + right[1], left[2] + right[2]


def state_at_time(history: Sequence[SourceState], time: float) -> SourceState:
    """Interpolate a trajectory with cubic Hermite positions."""

    if not history:
        raise ValueError("trajectory history must not be empty")
    first, last = history[0], history[-1]
    if time <= first.t:
        return replace(first, t=time, vx=0.0, vy=0.0, vz=0.0, ax=0.0, ay=0.0, az=0.0)
    if time >= last.t:
        return replace(last, t=time)

    lower, upper = 0, len(history) - 1
    while upper - lower > 1:
        middle = (lower + upper) // 2
        if history[middle].t <= time:
            lower = middle
        else:
            upper = middle

    left, right = history[lower], history[upper]
    duration = right.t - left.t
    amount = (time - left.t) / duration if duration > EPSILON else 0.0
    amount2, amount3 = amount**2, amount**3
    h00 = 2 * amount3 - 3 * amount2 + 1
    h10 = amount3 - 2 * amount2 + amount
    h01 = -2 * amount3 + 3 * amount2
    h11 = amount3 - amount2

    def position(name: str, velocity_name: str) -> float:
        return (
            h00 * getattr(left, name)
            + h10 * duration * getattr(left, velocity_name)
            + h01 * getattr(right, name)
            + h11 * duration * getattr(right, velocity_name)
        )

    return SourceState(
        t=time,
        x=position("x", "vx"),
        y=position("y", "vy"),
        z=position("z", "vz"),
        vx=_lerp(left.vx, right.vx, amount),
        vy=_lerp(left.vy, right.vy, amount),
        vz=_lerp(left.vz, right.vz, amount),
        ax=_lerp(left.ax, right.ax, amount),
        ay=_lerp(left.ay, right.ay, amount),
        az=_lerp(left.az, right.az, amount),
    )


def _arrival_residual(
    point: Vector3,
    observation_time: float,
    history: Sequence[SourceState],
    propagation_speed: float,
    emission_time: float,
) -> float:
    source = state_at_time(history, emission_time)
    distance = math.dist(point, (source.x, source.y, source.z))
    return emission_time + distance / propagation_speed - observation_time


def retarded_state(
    point: Vector3,
    observation_time: float,
    history: Sequence[SourceState],
    propagation_speed: float,
) -> SourceState:
    """Return the source state on the observation event's past light cone."""

    if propagation_speed <= 0:
        raise ValueError("propagation speed must be positive")
    if not history:
        raise ValueError("trajectory history must not be empty")

    first, last = history[0], history[-1]
    if _arrival_residual(point, observation_time, history, propagation_speed, first.t) >= 0:
        distance = math.dist(point, (first.x, first.y, first.z))
        return replace(
            first,
            t=observation_time - distance / propagation_speed,
            vx=0.0,
            vy=0.0,
            vz=0.0,
            ax=0.0,
            ay=0.0,
            az=0.0,
        )
    if _arrival_residual(point, observation_time, history, propagation_speed, last.t) <= 0:
        distance = math.dist(point, (last.x, last.y, last.z))
        return replace(
            last,
            t=observation_time - distance / propagation_speed,
            vx=0.0,
            vy=0.0,
            vz=0.0,
            ax=0.0,
            ay=0.0,
            az=0.0,
        )

    lower, upper = first.t, last.t
    for _ in range(48):
        middle = (lower + upper) / 2
        residual = _arrival_residual(point, observation_time, history, propagation_speed, middle)
        if residual > 0:
            upper = middle
        else:
            lower = middle
    return state_at_time(history, (lower + upper) / 2)


def lienard_wiechert_field(
    point: Vector3,
    observation_time: float,
    history: Sequence[SourceState],
    *,
    propagation_speed: float = 3.0,
    charge: float = 1.0,
    coulomb_constant: float = 1.0,
    softening: float = 0.12,
) -> FieldSample:
    """Evaluate velocity and radiation terms of the retarded fields."""

    source = retarded_state(point, observation_time, history, propagation_speed)
    displacement = (
        point[0] - source.x,
        point[1] - source.y,
        point[2] - source.z,
    )
    geometric_distance = math.sqrt(sum(component**2 for component in displacement))
    zero = (0.0, 0.0, 0.0)
    if geometric_distance < EPSILON:
        return FieldSample(zero, zero, zero, zero, zero, zero, source, source.t, 0.0)

    distance = max(softening, geometric_distance)
    direction = _scale(displacement, 1 / geometric_distance)
    beta = (
        source.vx / propagation_speed,
        source.vy / propagation_speed,
        source.vz / propagation_speed,
    )
    beta_squared = min(0.98, sum(component**2 for component in beta))
    beta_dot = (
        source.ax / propagation_speed,
        source.ay / propagation_speed,
        source.az / propagation_speed,
    )
    kappa = max(0.035, 1 - sum(a * b for a, b in zip(direction, beta, strict=True)))
    n_minus_beta: Vector3 = (
        direction[0] - beta[0],
        direction[1] - beta[1],
        direction[2] - beta[2],
    )
    velocity_factor = charge * coulomb_constant * (1 - beta_squared) / (kappa**3 * distance**2)
    velocity_electric = _scale(n_minus_beta, velocity_factor)

    n_dot_beta_dot = sum(a * b for a, b in zip(direction, beta_dot, strict=True))
    n_dot_n_minus_beta = sum(a * b for a, b in zip(direction, n_minus_beta, strict=True))
    radiation_numerator: Vector3 = (
        n_minus_beta[0] * n_dot_beta_dot - beta_dot[0] * n_dot_n_minus_beta,
        n_minus_beta[1] * n_dot_beta_dot - beta_dot[1] * n_dot_n_minus_beta,
        n_minus_beta[2] * n_dot_beta_dot - beta_dot[2] * n_dot_n_minus_beta,
    )
    radiation_factor = charge * coulomb_constant / (propagation_speed * kappa**3 * distance)
    radiation_electric = _scale(radiation_numerator, radiation_factor)
    electric = _add(velocity_electric, radiation_electric)
    velocity_magnetic = _scale(_cross(direction, velocity_electric), 1 / propagation_speed)
    radiation_magnetic = _scale(_cross(direction, radiation_electric), 1 / propagation_speed)
    magnetic = _add(velocity_magnetic, radiation_magnetic)
    return FieldSample(
        electric,
        magnetic,
        velocity_electric,
        radiation_electric,
        velocity_magnetic,
        radiation_magnetic,
        source,
        source.t,
        geometric_distance,
    )


def field_magnitude(vector: Vector3) -> float:
    """Return a three-vector's Euclidean norm."""

    return math.sqrt(sum(component**2 for component in vector))


def larmor_power(charge: float, acceleration: Vector3, propagation_speed: float) -> float:
    """Return the nonrelativistic Larmor power in the chosen units."""

    if propagation_speed <= 0:
        raise ValueError("propagation speed must be positive")
    return (
        2 * charge**2 * sum(component**2 for component in acceleration) / (3 * propagation_speed**3)
    )

File: visualizations/moving-charge-radiation/static/physics.js
const EPSILON = 1e-12;

function lerp(left, right, amount) {
  return left + (right - left) * amount;
}

function toVector3(vector) {
  return {x: vector.x ?? 0, y: vector.y ?? 0, z: vector.z ?? 0};
}

function cross(left, right) {
  return {
    x: left.y * right.z - left.z * right.y,
    y: left.z * right.x - left.x * right.z,
    z: left.x * right.y - left.y * right.x,
  };
}

export function stateAtTime(history, time) {
  if (!history.length) throw new Error("trajectory history must not be empty");
  const first = history[0];
  const last = history.at(-1);
  if (time <= first.t) return {...first, t: time, vx: 0, vy: 0, ax: 0, ay: 0};
  if (time >= last.t) return {...last, t: time};

  let lower = 0;
  let upper = history.length - 1;
  while (upper - lower > 1) {
    const middle = Math.floor((lower + upper) / 2);
    if (history[middle].t <= time) lower = middle;
    else upper = middle;
  }

  const left = history[lower];
  const right = history[upper];
  const duration = right.t - left.t;
  const u = duration > EPSILON ? (time - left.t) / duration : 0;
  const u2 = u * u;
  const u3 = u2 * u;
  const h00 = 2 * u3 - 3 * u2 + 1;
  const h10 = u3 - 2 * u2 + u;
  const h01 = -2 * u3 + 3 * u2;
  const h11 = u3 - u2;

  return {
    t: time,
    x: h00 * left.x + h10 * duration * left.vx + h01 * right.x + h11 * duration * right.vx,
    y: h00 * left.y + h10 * duration * left.vy + h01 * right.y + h11 * duration * right.vy,
    vx: lerp(left.vx, right.vx, u),
    vy: lerp(left.vy, right.vy, u),
    ax: lerp(left.ax, right.ax, u),
    ay: lerp(left.ay, right.ay, u),
  };
}

function arrivalResidual(point, observationTime, history, propagationSpeed, emissionTime) {
  const source = stateAtTime(history, emissionTime);
  const distance = Math.hypot(
    point.x - source.x,
    point.y - source.y,
    (point.z ?? 0) - (source.z ?? 0),
  );
  return emissionTime + distance / propagationSpeed - observationTime;
}

export function retardedState(point, observationTime, history, propagationSpeed) {
  if (!(propagationSpeed > 0)) throw new Error("propagation speed must be positive");
  if (!history.length) throw new Error("trajectory history must not be empty");

  const first = history[0];
  const last = history.at(-1);
  const firstResidual = arrivalResidual(
    point, observationTime, history, propagationSpeed, first.t,
  );
  if (firstResidual >= 0) {
    const distance = Math.hypot(
      point.x - first.x,
      point.y - first.y,
      (point.z ?? 0) - (first.z ?? 0),
    );
    return {...first, t: observationTime - distance / propagationSpeed, vx: 0, vy: 0, ax: 0, ay: 0};
  }

  const lastResidual = arrivalResidual(
    point, observationTime, history, propagationSpeed, last.t,
  );
  if (lastResidual <= 0) {
    const distance = Math.hypot(
      point.x - last.x,
      point.y - last.y,
      (point.z ?? 0) - (last.z ?? 0),
    );
    return {...last, t: observationTime - distance / propagationSpeed, vx: 0, vy: 0, ax: 0, ay: 0};
  }

  let lower = first.t;
  let upper = last.t;
  for (let iteration = 0; iteration < 32; iteration += 1) {
    const middle = (lower + upper) / 2;
    const residual = arrivalResidual(
      point, observationTime, history, propagationSpeed, middle,
    );
    if (residual > 0) upper = middle;
    else lower = middle;
  }
  return stateAtTime(history, (lower + upper) / 2);
}

function zeroField(source) {
  return {
    electric: {x: 0, y: 0, z: 0},
    magnetic: {x: 0, y: 0, z: 0},
    magneticZ: 0,
    velocityElectric: {x: 0, y: 0, z: 0},
    radiationElectric: {x: 0, y: 0, z: 0},
    velocityMagnetic: {x: 0, y: 0, z: 0},
    radiationMagnetic: {x: 0, y: 0, z: 0},
    velocityMagneticZ: 0,
    radiationMagneticZ: 0,
    source,
    retardedTime: source.t,
    distance: 0,
  };
}

export function lienardWiechertField(point, observationTime, history, options = {}) {
  const propagationSpeed = options.propagationSpeed ?? 3;
  const charge = options.charge ?? 1;
  const coulombConstant = options.coulombConstant ?? 1;
  const softening = options.softening ?? 0.12;
  const source = retardedState(point, observationTime, history, propagationSpeed);
  const observation = toVector3(point);
  const sourcePosition = toVector3(source);
  const displacement = {
    x: observation.x - sourcePosition.x,
    y: observation.y - sourcePosition.y,
    z: observation.z - sourcePosition.z,
  };
  const geometricDistance = Math.hypot(
    displacement.x,
    displacement.y,
    displacement.z,
  );
  if (geometricDistance < EPSILON) return zeroField(source);

  const distance = Math.max(softening, geometricDistance);
  const direction = {
    x: displacement.x / geometricDistance,
    y: displacement.y / geometricDistance,
    z: displacement.z / geometricDistance,
  };
  const beta = {
    x: source.vx / propagationSpeed,
    y: source.vy / propagationSpeed,
    z: (source.vz ?? 0) / propagationSpeed,
  };
  const betaSquared = Math.min(
    0.98,
    beta.x * beta.x + beta.y * beta.y + beta.z * beta.z,
  );
  const betaDot = {
    x: source.ax / propagationSpeed,
    y: source.ay / propagationSpeed,
    z: (source.az ?? 0) / propagationSpeed,
  };
  const kappa = Math.max(
    0.035,
    1 - direction.x * beta.x - direction.y * beta.y - direction.z * beta.z,
  );
  const denominator = kappa ** 3;
  const nMinusBeta = {
    x: direction.x - beta.x,
    y: direction.y - beta.y,
    z: direction.z - beta.z,
  };
  const velocityFactor = charge * coulombConstant * (1 - betaSquared)
    / (denominator * distance ** 2);
  const velocityElectric = {
    x: velocityFactor * nMinusBeta.x,
    y: velocityFactor * nMinusBeta.y,
    z: velocityFactor * nMinusBeta.z,
  };

  // n × ((n − β) × βdot), evaluated with the vector triple-product identity.
  const nDotBetaDot = direction.x * betaDot.x
    + direction.y * betaDot.y
    + direction.z * betaDot.z;
  const nDotNMinusBeta = direction.x * nMinusBeta.x
    + direction.y * nMinusBeta.y
    + direction.z * nMinusBeta.z;
  const radiationNumerator = {
    x: nMinusBeta.x * nDotBetaDot - betaDot.x * nDotNMinusBeta,
    y: nMinusBeta.y * nDotBetaDot - betaDot.y * nDotNMinusBeta,
    z: nMinusBeta.z * nDotBetaDot - betaDot.z * nDotNMinusBeta,
  };
  const radiationFactor = charge * coulombConstant
    / (propagationSpeed * denominator * distance);
  const radiationElectric = {
    x: radiationFactor * radiationNumerator.x,
    y: radiationFactor * radiationNumerator.y,
    z: radiationFactor * radiationNumerator.z,
  };
  const electric = {
    x: velocityElectric.x + radiationElectric.x,
    y: velocityElectric.y + radiationElectric.y,
    z: velocityElectric.z + radiationElectric.z,
  };
  const velocityMagneticCross = cross(direction, velocityElectric);
  const radiationMagneticCross = cross(direction, radiationElectric);
  const velocityMagnetic = {
    x: velocityMagneticCross.x / propagationSpeed,
    y: velocityMagneticCross.y / propagationSpeed,
    z: velocityMagneticCross.z / propagationSpeed,
  };
  const radiationMagnetic = {
    x: radiationMagneticCross.x / propagationSpeed,
    y: radiationMagneticCross.y / propagationSpeed,
    z: radiationMagneticCross.z / propagationSpeed,
  };
  const magnetic = {
    x: velocityMagnetic.x + radiationMagnetic.x,
    y: velocityMagnetic.y + radiationMagnetic.y,
    z: velocityMagnetic.z + radiationMagnetic.z,
  };

  return {
    electric,
    magnetic,
    magneticZ: magnetic.z,
    velocityElectric,
    radiationElectric,
    velocityMagnetic,
    radiationMagnetic,
    velocityMagneticZ: velocityMagnetic.z,
    radiationMagneticZ: radiationMagnetic.z,
    source,
    retardedTime: source.t,
    distance: geometricDistance,
  };
}

export function fieldMagnitude(vector) {
  return Math.hypot(vector.x, vector.y, vector.z ?? 0);
}

export function clampVector(vector, maximum) {
  const magnitude = fieldMagnitude(vector);
  if (magnitude <= maximum || magnitude < EPSILON) return {...vector};
  const scale = maximum / magnitude;
  return {x: vector.x * scale, y: vector.y * scale};
}

export function larmorPower(charge, acceleration, propagationSpeed) {
  const accelerationSquared = acceleration.x ** 2 + acceleration.y ** 2;
  return 2 * charge ** 2 * accelerationSquared / (3 * propagationSpeed ** 3);
}
